addons/binding/org.openhab.binding.nanoleaf/src/main/java/org/openhab/binding/nanoleaf/internal/handler/NanoleafPanelHandler.java

```diff
@@ -0,0 +1,225 @@
+import org.openhab.binding.nanoleaf.internal.config.NanoleafControllerConfig;
+import org.openhab.binding.nanoleaf.internal.config.NanoleafPanelConfig;
+import org.openhab.binding.nanoleaf.internal.model.Effects;
+import org.openhab.binding.nanoleaf.internal.model.Write;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link NanoleafPanelHandler} is responsible for handling commands to the controller which
+ * affect an individual panels
+ *
+ * @author Martin Raepple - Initial contribution
+ */
+public class NanoleafPanelHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(NanoleafPanelHandler.class);
+    private HttpClient httpClient;
+    // JSON parser for API responses
+    private final Gson gson = new Gson();
+
+    public NanoleafPanelHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("initializing handler for panel {}", getThing().getUID());
+        updateStatus(ThingStatus.OFFLINE);
+        Bridge controller = getBridge();
+        if (controller == null) {
+            initializePanel(null, null);
+        } else {
+            initializePanel(controller.getHandler(), controller.getStatus());
+        }
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo controllerStatusInfo) {
+        logger.debug("Controller status changed to {}", controllerStatusInfo);
+        Bridge controller = getBridge();
+        if (controller == null) {
+            initializePanel(null, controllerStatusInfo.getStatus());
+        } else {
+            initializePanel(controller.getHandler(), controllerStatusInfo.getStatus());
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Received command {} for channel {}", command, channelUID);
+        try {
+            switch (channelUID.getId()) {
+                case CHANNEL_PANEL_COLOR:
+                    sendRenderedEffectCommand(command);
+                    break;
+                case CHANNEL_PANEL_BRIGHTNESS:
+                    sendRenderedEffectCommand(command);
+                    break;
+                default:
+                    logger.warn("Channel with id {} not handled", channelUID.getId());
+                    break;
+            }
+        } catch (NanoleafUnauthorizedException nae) {
+            logger.warn("Authorization for command {} for channelUID {} failed: {}", command, channelUID,
+                    nae.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Invalid token. Replace with valid token or start pairing again.");
+            NanoleafControllerConfig controllerConfig = getConfigAs(NanoleafControllerConfig.class);
```

Comment:
Yes, it is the contoller configuration, which is checked here for the authorization token. If none is present, the pairing job is started on the controller handler.

Reply:
I would be surprised if this works. Because the controller config is bound to the bridge not this `NanoleafPanelHandler` given the xml definitions of the thing. While it technically probably will work. I assume it tries to find match the fields with the properties and while find none matching because it contains different properties. and will have the default values set in the class. But I would not expect it to get the actual values of the controller. So the next line below with always return true.

Reply:
I see - so I'd suggest to remove the code which tries to restart the pairingJob from the `NanoleafPanelHandler` here. From a user's perspective, it should be sufficient to set the panel ot OFFLINE status, which will anyways happen on the controller/bridge as well in case something is wrong with the authorization token. The error message then instructs the user to either provide a valid token, or remove it from the controller config to start the pairing process.

Reply:
Usage of `NanoleafControllerConfig` is now removed from the `NanoleafPanelHandler`